Prototype MVC changes for next auth iteration

HaoK opened the pull request:
- Prototype switch to using common AuthorizeAttribute, still need to rationalize the AuthorizationFilterAttribute, this change focused on removing the MVC AuthorizeAttribute and preserving the exact same functionality (tests all pass)

Should what I call AuthorizationFilter, really be called AuthorizeFilter since it basically is a filter that was created from an Authorize attribute?

cc @yishaigalatzer @harshgMSFT @rynowak @blowdart @lodejard

blowdart commented:
Naming is hard. But yea, AuthorizeFilter seems better.

yishaigalatzer commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizationFilter.cs:
```diff
@@ -1,46 +1,25 @@
 {
-    public class AuthorizeAttribute : AuthorizationFilterAttribute
+    // REVIEW: figure out how to propery rationalize this with AuthorizationFilterAttribute
+    public class AuthorizationFilter : IAsyncAuthorizationFilter, IAuthorizationFilter, IOrderedFilter
```
you don't need to implement both AsyncAuthorizationFilter and `IAuthorizationFilter` it was there for overriding the property but probably no longer necessary?

HaoK replied:
Removed sync version, only implements AsyncAuthorizationFilter now

yishaigalatzer commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizationFilter.cs:
```diff
@@ -53,11 +32,11 @@ public string Roles
                 var rolesPolicy = new AuthorizationPolicyBuilder();
-                rolesPolicy.RequiresRole(_rolesSplit);
-                if (!await authService.AuthorizeAsync(rolesPolicy.Build(), httpContext, context))
+                rolesPolicy.RequiresRole(Roles.ToArray());
```
Lets not `ToArray` on every call

HaoK replied:
Fair enough, I'll add an overload to Policy builder which also allows `IEnumerable<string>`

yishaigalatzer commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,72 @@
+
+        public string Policy { get; set; }
+
+        public IEnumerable<string> Roles { get; set; }
```
Make this an `IList<string>` so no ToArray is required later.

HaoK replied:
The issue is that the AuthZ apis only took params string[], so IList doesn't help, I added the overloads that accept IEnumerable

rynowak replied:
I don't really grok this comment or the response. I don't see a `ToArray()` so I'm assuming there's no issue here?

HaoK replied:
Yeah I fixed it in security itself by adding overloads that take IEnumerable directly, so there no longer is an issue with ToArray

rynowak commented on src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs:
```diff
@@ -256,6 +257,15 @@ private bool IsIDisposableMethod(MethodInfo methodInfo, TypeInfo typeInfo)
+                    Policy = authorizeAttribute.Policy,
+                    Roles = authorizeAttribute.Roles?.Split(','),
+                });
+            }
```
We need to see the same in `ControllerModelBuilder`. This code only looks at attributes on the method.

rynowak replied:
Need to see a unit test for this and also for controllermodelbuilder when that's fixed too.

HaoK replied:
Added unit tests

rynowak replied:
:+1:

rynowak commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,72 @@
+
+        public IEnumerable<string> Roles { get; set; }
+
+#pragma warning disable 1998
```
why? looks like this method does have async code.

HaoK replied:
This was the old AuthorizeAttribute code, I just copied it over, I can remove it

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs:
```diff
@@ -1835,15 +1835,15 @@ private struct SampleStruct
                 });
 
-            var invoker = CreateInvoker(new IFilter[] { authorizationFilter.Object, resourceFilter.Object, });
+            var invoker = CreateInvoker(new IFilter[] { AuthorizeFilter.Object, resourceFilter.Object, });
```
revert these - find and replace fail

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/DefaultActionDescriptorCollectionProviderLoggingTest.cs:
```diff
@@ -72,7 +73,7 @@ public void ControllerDiscovery()
             Assert.Empty(controllerModelValues.RouteConstraints);
             Assert.NotEmpty(controllerModelValues.Attributes);
-            Assert.Single(controllerModelValues.Filters);
+            Assert.Empty(controllerModelValues.Filters);
```
This shouldn't have changed - fix the `ControllerModelBuilder` and this will be the same as before

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeAttributeTests.cs:
```diff
@@ -15,7 +15,7 @@ public class AuthorizeAttributeTests : AuthorizeAttributeTestsBase
         public async Task Invoke_ValidClaimShouldNotFail()
```
rename file

HaoK replied:
Renamed

rynowak commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,72 @@
+namespace Microsoft.AspNet.Mvc
+{
+    // REVIEW: figure out how to propery rationalize this with AuthorizationFilterAttribute
+    public class AuthorizeFilter : IAsyncAuthorizationFilter, IOrderedFilter
```
needs docs

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs:
```diff
@@ -321,6 +322,27 @@ public void GetActions_ConventionallyRoutedAction_WithHttpConstraints()
         }
 
         [Fact]
+        public void GetActions_AuthorizeFiltersAreCombinedWhenOverriden()
```
This isn't really accurate is it? 2 filters are created and both with run so how are they combined?

HaoK replied:
Poor test name, the intent was that both base and derived authorize filters show up, as opposed to getting only one.  I'll rename it to BaseAuthorizeFiltersAreStillValidWhenOverriden

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultControllerModelBuilderTest.cs:
```diff
@@ -7,6 +7,8 @@
 using System;
 using System.Threading.Tasks;
+using Microsoft.AspNet.Security;
+using System.Linq;
```
sort

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTests.cs:
```diff
@@ -9,13 +9,13 @@
 
 namespace Microsoft.AspNet.Mvc.Core.Test
```
namespace should match product code

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTestsBase.cs:
```diff
@@ -12,7 +12,7 @@
-    public class AuthorizeAttributeTestsBase
+    public class AuthorizeFilterTestsBase
     {
         protected AuthorizationContext GetAuthorizationContext(Action<ServiceCollection> registerServices, bool anonymous = false)
```
This is wierd and only used in one place. I know it's not a mess that you made but you can just move this method to the other class and delete this?

rynowak commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,72 @@
+    /// <summary>
+    /// An implementation of <see cref="IAsyncAuthorizationFilter"/> which corresponds to [Authorize]
+    /// </summary>
+    public class AuthorizeFilter : IAsyncAuthorizationFilter, IOrderedFilter
```
Sorry, wasn't clear, need docs on all of the public/protected stuff. If it's an interface method you can just use `/// <inheritdoc />`

rynowak commented on test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTests.cs:
```diff
@@ -9,13 +9,13 @@
 namespace Microsoft.AspNet.Mvc.Core.Test
 {
-    public class AuthorizeAttributeTests : AuthorizeAttributeTestsBase
+    public class AuthorizeFilterTests : AuthorizeFilterTestsBase
```
`AuthorizeFilterTest` (yeah I know it was wrong before too)

rynowak commented:
:watch: looking pretty close to done

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,86 @@
+        public int Order { get; set; }
+
+        /// <summary>
+        /// Policy name used for authorization
```
nit: all summary should end with a .

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        /// <summary>
+        /// Returns true if there is an IAllowAnonymous filter
```
<see cref="IAllowAnonymous" />

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <inheritdoc />
+        protected virtual void Fail([NotNull] AuthorizationContext context)
+        {
+            context.Result = new HttpStatusCodeResult(401);
```
Is this the recommended way? should there be a Fail and Success on the service/context itself? and the authorization gets to 401 or 403 depending on what failed? I know this is doing exactly what the code before does, but wondering if the pattern is going to be this or going to change.

harshgMSFT replied:
also use ```Microsoft.AspNet.WebUtilities.StatusCodes.Status401Unauthorized ```  instead of 401

harshgMSFT commented on test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTest.cs:
```diff
@@ -1,21 +1,75 @@
-    public class AuthorizeAttributeTests : AuthorizeAttributeTestsBase
+    public class AuthorizeFilterTest
     {
+        private AuthorizationContext GetAuthorizationContext(Action<ServiceCollection> registerServices, bool anonymous = false)
```
nit: can we move this to end of the file? that is what we have been following.

harshgMSFT commented:
:watch:

HaoK commented:
Updated with latest iteration changes, and PR feedback:
- AuthorizeFilter expects to get a combined policy (union of all the authorize attributes)
- Returns 401 if anonymous, 403 if authorization service fails
- Updated functional to be less verby

borgdylan commented:
Shoudldn't the AuthorizeFilter support a policy less constructor? That would make basic usage easier rather than having to define policies in DI for a simple case.

HaoK commented:
@borgdylan The new authorize filter expects the combined policies, if there is not supposed to be any authorization done at all (which is the empty policy less constructor case) there shouldn't be an authorize filter at all.  This is also a side effect of the decision to make an empty policy with no requirements always fail auth.

borgdylan commented:
But will there be a way to make a policy that allows anyone that presents valid credentials to pass successfully?

HaoK commented:
Sure, policy.RequireAnyAuthenticatedUser()

HaoK commented:
@harshgMSFT @yishaigalatzer @rynowak Is this PR ready for checkin?  There's another big wave of authN changes coming, so I want to get this iteration of AuthZ in before tackling that one.

lodejard commented:
:shipit: ..?

harshgMSFT commented on samples/MvcSample.Web/Startup.cs:
```diff
@@ -107,6 +107,19 @@ public void Configure(IApplicationBuilder app)
                     });
+                    services.ConfigureAuthorization(auth =>
+                    {
+                        auth.AddPolicy("CanViewPage",
```
nit: formatting : all parameters should be on top of one another. 
something like 
```c#
    auth.AddPolicy(
        "CanViewAnything",
        new AuthorizationPolicyBuilder().RequiresClaim("Permission", "CanViewAnything")
            .Build());
```

harshgMSFT commented on samples/MvcSample.Web/Startup.cs:
```diff
@@ -107,6 +107,19 @@ public void Configure(IApplicationBuilder app)
+                        auth.AddPolicy("CanViewAnything",
+                            new AuthorizationPolicyBuilder()
+                                .RequiresClaim("Permission", "CanViewAnything").Build());
+                        // This policy basically requires that the auth type is present
```
nit: blank line above this

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs:
```diff
@@ -16,6 +18,13 @@ namespace Microsoft.AspNet.Mvc.ApplicationModels
     {
+        private readonly AuthorizationOptions _authorizationOptions;
+
+        public DefaultActionModelBuilder(IOptions<AuthorizationOptions> options = null)
```
docs. And we do not use default values. Why is this needed?

HaoK replied:
Made it so I didn't have to update all the unit tests, but I can remove it

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs:
```diff
@@ -20,15 +22,17 @@ public class DefaultControllerModelBuilder : IControllerModelBuilder
         /// </summary>
         /// <param name="actionModelBuilder">The <see cref="IActionModelBuilder"/> used to create actions.</param>
-        public DefaultControllerModelBuilder(IActionModelBuilder actionModelBuilder, ILoggerFactory loggerFactory)
+        public DefaultControllerModelBuilder(IActionModelBuilder actionModelBuilder, ILoggerFactory loggerFactory, IOptions<AuthorizationOptions> options = null)
```
break the line so that length is <120

harshgMSFT replied:
as above, do not use default parameters.

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,58 @@
+    public class AuthorizeFilter : IAsyncAuthorizationFilter
+    {
+        /// <summary>
+        /// Authorize filter for a specific policy
```
All the docs should finish with a full stop ```.```

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <summary>
+        /// Authorize filter for a specific policy
+        /// </summary>
+        /// <param name="policy"></param>
```
missing docs.

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,58 @@
+            // Build a ClaimsPrincipal with the Policy's required authentication types
+            if (Policy.ActiveAuthenticationTypes != null && Policy.ActiveAuthenticationTypes.Any())
+            {
+                var results = await context.HttpContext.AuthenticateAsync(Policy.ActiveAuthenticationTypes);
```
can the results be null?

HaoK replied:
Added null guard

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,58 @@
+
+            var httpContext = context.HttpContext;
+            var authService = httpContext.RequestServices.GetRequiredService<IAuthorizationService>();
+            // Note: Default Anonymous User is new ClaimsPrincipal(new ClaimsIdentity())
```
nit: a newline above this.

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,58 @@
+            }
+
+            var httpContext = context.HttpContext;
+            var authService = httpContext.RequestServices.GetRequiredService<IAuthorizationService>();
```
Instead of calling getRequired service here(i.e within the filter), the pattern that we follow is to have an implementation of ```IFilterFactory``` and create the instance of the filter there. So in this case we would have something like 
```c# 
public class AuthorizeFilterFactory : IFilterFactory

```
and then create the filter in its create method ( the advantage here is that create method takes in an IServiceProvider which can be used to resolve the authorization service. ).

HaoK replied:
Can you point me to an example in the current sources?

harshgMSFT replied:
look at https://github.com/aspnet/Mvc/blob/b98de5c5a305718487f616a4f5f080fb2a59be78/src/Microsoft.AspNet.Mvc.Core/Filters/ResponseCacheAttribute.cs#L88

yishaigalatzer replied:
This doesn't add much value. I'm suggesting won't fix

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs:
```diff
@@ -126,6 +130,12 @@ protected virtual ControllerModel CreateControllerModel([NotNull] TypeInfo typeI
+            var policy = AuthorizationPolicy.Combine(_authorizationOptions, attributes.OfType<AuthorizeAttribute>());
+            if (policy != null)
+            {
+                controllerModel.Filters.Add(new AuthorizeFilter(policy));
```
Instead of creating the AuthorizeFilter here, probably we should create a filter factory. That would automatically create the filter at a later point.  Also creating the factory here will enable you to not take here in ```DefaultControllerModelBuilder``` and ```DefaultActionModelBuilder```. Please see the comment on AuthorizeFilter as well.

HaoK replied:
Filed https://github.com/aspnet/Mvc/issues/2022 to track this for later

rynowak replied:
I don't understand the comment here. `IFilterFactory` is only useful when when a filter needs indirection (replaceable) or needs access to request services. 

If you made this `IFilterFactory` you'd still need code here because `AuthorizeAttribute` cannot implement `IFilterFactory` itself.

harshgMSFT replied:
the filter factory is useful since the AuthorizeFilter can directly get the service via a factory. Agreed that this part of the code will still have to do what its doing today.

harshgMSFT commented on test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs:
```diff
@@ -646,8 +672,24 @@ public void GetActions_MixedHttpVerbsAndRoutes_MultipleEmptyAndNonEmptyVerbs()
             Assert.Equal<string>(new string[] { "GET" }, action.HttpMethods);
         }
 
+        private static DefaultActionModelBuilder CreateTestDefaultActionModelBuilder(AuthorizationOptions authOptions = null)
```
line breaks.

harshgMSFT commented on test/WebSites/XmlFormattersWebSite/XmlFormattersWebSite.xproj:
```diff
@@ -0,0 +1,20 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
revert.

harshgMSFT commented on test/WebSites/FiltersWebSite/Startup.cs:
```diff
@@ -9,6 +10,25 @@
+
+    public class ManagerHandler : AuthorizationHandler<OperationAuthorizationRequirement>
+    {
+        public override Task HandleAsync(Microsoft.AspNet.Security.AuthorizationContext context, OperationAuthorizationRequirement requirement)
```
line breaks.

harshgMSFT commented on test/WebSites/FiltersWebSite/Startup.cs:
```diff
@@ -9,6 +10,25 @@
+            {
+                context.Succeed(requirement);
+            }
+            return Task.FromResult(0);
```
nit: use ```Task.FromResult(true);```

harshgMSFT commented on test/WebSites/FiltersWebSite/Startup.cs:
```diff
@@ -9,6 +10,25 @@
+    {
+        public override Task HandleAsync(Microsoft.AspNet.Security.AuthorizationContext context, OperationAuthorizationRequirement requirement)
+        {
+            if (context.User.HasClaim("Manager", "yes"))
```
nit: named parameters. Will be easier to review and read.

harshgMSFT commented on test/WebSites/FiltersWebSite/AuthorizeBasicMiddleware.cs:
```diff
@@ -26,9 +25,12 @@ public class AuthorizeBasicMiddleware : AuthenticationMiddleware<BasicOptions>
-        { }
+            IOptions<BasicOptions> options,
+            string authType) : 
+                base(next, services, options, 
```
if you are splitting the parameters, have all of them on a newline.

harshgMSFT commented on test/WebSites/FiltersWebSite/AuthorizeBasicMiddleware.cs:
```diff
@@ -15,10 +16,8 @@ public class BasicOptions : AuthenticationOptions
     {
```
This file contains 3 classes, which is something we do not follow. (even thought this is not directly related to your change ) could you please split this file into 3?

harshgMSFT commented on test/WebSites/FiltersWebSite/Startup.cs:
```diff
@@ -9,6 +10,25 @@
 
 namespace FiltersWebSite
 {
+    public static class Operations
```
separate files for separate classes.

harshgMSFT commented on test/WebSites/FiltersWebSite/AuthorizeBasicMiddleware.cs:
```diff
@@ -49,12 +51,13 @@ protected override void ApplyResponseGrant()
         protected override AuthenticationTicket AuthenticateCore()
```
should this class be worried about handling a response challenge?

harshgMSFT commented:
:watch:

yishaigalatzer commented on src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs:
```diff
@@ -0,0 +1,62 @@
+                var results = await context.HttpContext.AuthenticateAsync(Policy.ActiveAuthenticationTypes);
+                if (results != null)
+                {
+                    context.HttpContext.User = new ClaimsPrincipal(results.Where(r => r.Identity != null).Select(r => r.Identity));
```
fix the long line

HaoK commented:
Merged, will address any leftover feedback with the next authZ/N iteration update 94ef6cf8fa9e62bca108ef20161f77ce45789a04
